routes/socket.js

```diff
@@ -1,133 +1,48 @@
-    io.emit('hello', 'to all clients');
-    // Simulate that the game-master started a game in a game room using
-
-    io
-      .to(`room-${roomId}`)
-      .emit('game:start', { roomId, message: 'Game-master has started the game' });
-
-    // Check which clients are connected to this room
-    io.in(`room-${roomId}`).clients((err, clients) => {
-      logger.info('Clients in room', { roomId, clients });
-    });
-  }, 10000);
-
-  // const interval = setInterval(() => {
-  //   console.log('Notifying all listeners');
-  //   io.emit('master:notify', { message: `Hello ${count}` })
-  //   count++;
-  // }, 3000);
-
-  socket.on('task:start', (task) => {
-    logger.info('task:start', { socketMessage: task, roomId });
-    socket.to(`room-${roomId}`).emit('task:start', { task });
-    room.update({
-      currentTaskId: task.id
-    });
-  });
-
-  socket.on('task:ending', () => {
-    logger.info('task:ending', { roomId });
-    socket.to(`room-${roomId}`).emit('task:ending');
-  });
-
-  socket.on('task:answer', (payload) => {
-    logger.info('task:answer', { socketMessage: payload, roomId, gameMasterSocketId });
-    const task = room.getCurrentTask();
-    player = playerService.incrementScoreIfAnswerCorrect(payload.answer, task, player);
-    io.to(gameMasterSocketId).emit('task:answer', { payload, player });
-  });
-
-  socket.on('round:ending', () => {
-    logger.info('round:ending', { roomId });
-    room.getPlayers().then((players) => {
-      players.forEach((playerObj) => {
-        io.to(playerObj.socket).emit('round:ending', { player: playerObj });
-      });
-    });
-  });
-
-  socket.on('game:start', (game) => {
-    logger.info('game:start', { socketMessage: game, roomId });
-    socket.to(`room-${roomId}`).emit('game:start', { game });
-  });
-
-  socket.on('game:ending', (game) => {
-    logger.info('game:ending', { socketMessage: game, roomId });
-    socket.to(`room-${roomId}`).emit('game:ending', { game });
-  });
-
-  socket.on('disconnect', () => {
-    logger.info('Client disconnected', { roomId });
-  });
-
-  socket.on('game message', (msg) => {
-    logger.info('Game message', { socketMessage: msg, roomId });
-  });
-};
-
 const initSocket = (server) => {
   const io = new SocketServer(server);
   io.adapter(redisAdapter({ host: process.env.REDIS_HOST, port: 6379 }));
 
   // Authenticate connection
   // Will only run once per client-server connection
   io.use((socket, next) => {
-    const { token } = socket.handshake.query;
-    getRoom(token).then((room) => {
-      logger.info('Authenticating client...', { token, room });
-      logger.info('RoomId', { room });
-      if (!room) {
-        logger.warn('Token provided by client was not valid', { token });
-        return next(new Error('Invalid joinToken'));
-      }
-      if (!room.gameMasterID) {
+    const { token, role } = socket.handshake.query;
+    if (typeof role !== 'undefined' && role === 'master') {
+      getRoom(token).then((room) => {
```

Comment:
Eg ser også at det er ein del duplikert kode i `if`- og `else`-blokka, og det som hovedsakeleg er forskjellen er:

```js
room.update({
  gameMasterID: socket.id
})
```
Eg foreslår at du heller gjenbruker all logikken du kan gjenbruke, og venter med å sjekke om det GM eller Player, til du trenger å skille dei. Då kan du bruke chaining av promises for å få til den logikken, då vil det også likne meir på eit pipe and filter mønster

```js
logger.info('Authenticating Client...', { token, room, role });
getRoom(token)
  .then((room) => {
    if (!room) {
      logger.warn('Token provided by client was not valid', { token });
      // Send feilen vidare til ein felles .catch blokk lenger nede
      return Promise.reject(new Error('Invalid joinToken'));
    }

    if (role === 'master') {
      room.update({
        gameMasterID: socket.id
      });
    }
    
    return next();
  })
 .catch((error) => next(error);
```

Alternativt kan du også gjere det slikt:

```js
logger.info('Authenticating Client...', { token, room, role });
getRoom(token)
  .then((room) => {
    if (!room) {
      logger.warn('Token provided by client was not valid', { token });
      // Send feilen vidare til ein felles .catch blokk lenger nede
      return Promise.reject(new Error('Invalid joinToken'));
    }

    return Promise.resolve(room);
  })
  .then((room) => {
    if (role === 'master') {
      room.update({
        gameMasterID: socket.id
      });
    }
    return next();
  })
 .catch((error) => next(error);
```